src/librustc_typeck/mem_categorization.rs

```diff
@@ -609,6 +640,58 @@ impl<'a, 'tcx> MemCategorizationContext<'a, 'tcx> {
         self.cat_pattern_(place, pat, &mut op)
     }
 
+    fn variant_index_for_adt(
```

Comment:
Can you add a comment here explaining what the parameters represent? In particular, what is this `hir_id` the id of?